pkg/disk/bdf.go

```diff
@@ -570,5 +631,39 @@ func (d *driver) UnbindDriver() error {
 }
 
 func (d *driver) BindDriver(targetDriver string) error {
-	return utilsio.WriteTrunc(unix.AT_FDCWD, filepath.Join(sysPrefix, "sys/bus", d.machineType.BusName(), "drivers", targetDriver, "bind"), []byte(d.deviceNumber))
+	err := utilsio.WriteTrunc(unix.AT_FDCWD, filepath.Join(sysPrefix, "sys/bus", d.machineType.BusName(), "devices", d.deviceNumber, "driver_override"), []byte(targetDriver))
+	if err != nil {
+		return err
+	}
+	if d.machineType == BDF {
+		return utilsio.WriteTrunc(unix.AT_FDCWD, filepath.Join(sysPrefix, "sys/bus", d.machineType.BusName(), "drivers_probe"), []byte(d.deviceNumber))
+	}
+	return nil
+}
+
+func (d *driver) GetPCIDeviceDriverType() string {
+	output, _ := exec.Command("lspci", "-s", d.deviceNumber, "-n").CombinedOutput()
+	klog.InfoS("GetDeviceDriverType: get driver type output", "deviceNumber", d.deviceNumber, "output", output)
+	// #define PCI_DEVICE_ID_VIRTIO_BLOCK 0x1001
+	// #define PCI_DEVICE_ID_ALIBABA_NVME 0×5004
+	if strings.HasSuffix(strings.TrimSpace(string(output)), "1001") {
+		return PCITypeVIRTIO
+	} else {
+		return PCITypeNVME
+	}
+}
+
+func (d *driver) CheckVFIOUsage() error {
+	actualPath, err := filepath.EvalSymlinks(filepath.Join(sysPrefix, "sys/bus", d.machineType.BusName(), "devices", d.deviceNumber, "iommu_group"))
+	if err != nil {
+		return err
+	}
+	klog.V(5).InfoS("CheckVFIOUsage: eval symlink success", "path", actualPath)
+	groupNumber := filepath.Base(actualPath)
+	// the command returns -1 if nothing is returned
+	output, _ := exec.Command("lsof", filepath.Join("/dev/vfio", groupNumber)).CombinedOutput()
```

Comment:
This is expensive, I would strongly recommend not using this. Instead, ask the user to lock this file, so that we can use lock to detect usage. Or ask the user to write a PID file, and we can just check if the process is alive.